packages/protobuf/src/next/types.ts

```diff
@@ -0,0 +1,70 @@
+// Copyright 2021-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type {
+  TypedDescEnum,
+  TypedDescMessage,
+} from "./codegenv1/typed-desc.js";
+import type { DescEnum, DescMessage } from "../descriptor-set.js";
+
+export type Message<TypeName extends string = string> = {
+  readonly $desc: DescMessage;
+  readonly $typeName: TypeName;
+
+  // TODO support unknown fields
+  // $unknown?: ({no: number; wireType: WireType; data: Uint8Array; })[];
+};
+
+// TODO docs
+export type MessageShape<Desc extends DescMessage> =
+  Desc extends TypedDescMessage<infer RuntimeShape> ? RuntimeShape : Message;
+
+// TODO docs
+export type MessageInitShape<Desc extends DescMessage> =
+  Desc extends TypedDescMessage<infer RuntimeShape>
+    ? RuntimeShape | MessageInit<RuntimeShape>
+    : Record<string, unknown>; // TODO better input type
+
+// TODO docs
+export type EnumShape<Desc extends DescEnum> =
+  Desc extends TypedDescEnum<infer RuntimeShape> ? RuntimeShape : number;
+
+// TODO ServiceShape
+// TODO MethodShape?
+
+type MessageInit<T extends Message> = {
```

Comment:
And it also makes all properties optional. It's a replacement for `PartialMessage` with a more fitting name.

Arguably, proto2 required fields should not be optional when creating a message (https://github.com/bufbuild/protobuf-es/issues/414). We can consider generating an "input shape" attached to `TypedDescMessage` for messages using proto2 required, and fall back to `MessageInit`. This might come with a lot of complication (also for users), and required fields are heavily discouraged. Best to punt on this decision for now.